index.d.ts

```diff
@@ -0,0 +1,83 @@
+declare function pipe <T1, T2, R1, R2, R3, R4, R5, R6, R>(f1:BinaryFunction<T1, T2, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R5>, f6:UnaryFunction<R5, R6>, f7:UnaryFunction<R6, R>): Pipeline<[T1, T2], R>;
+declare function pipe <T1, T2, R1, R2, R3, R4, R5, R6, R7, R>(f1:BinaryFunction<T1, T2, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R5>, f6:UnaryFunction<R5, R6>, f7:UnaryFunction<R6, R7>, f8:UnaryFunction<R7, R>): Pipeline<[T1, T2], R>;
+declare function pipe <T1, T2, R1, R2, R3, R4, R5, R6, R7, R8, R>(f1:BinaryFunction<T1, T2, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R5>, f6:UnaryFunction<R5, R6>, f7:UnaryFunction<R6, R7>, f8:UnaryFunction<R7, R8>, f9:UnaryFunction<R8, R>): Pipeline<[T1, T2], R>;
+
+declare function pipe <T1, T2, T3, R>(f1:TernaryFunction<T1, T2, T3, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R2, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R2, R3, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R2, R3, R4, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R2, R3, R4, R5, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R5>, f6:UnaryFunction<R5, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R2, R3, R4, R5, R6, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R5>, f6:UnaryFunction<R5, R6>, f7:UnaryFunction<R6, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R2, R3, R4, R5, R6, R7, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R5>, f6:UnaryFunction<R5, R6>, f7:UnaryFunction<R6, R7>, f8:UnaryFunction<R7, R>): Pipeline<[T1, T2, T3], R>;
+declare function pipe <T1, T2, T3, R1, R2, R3, R4, R5, R6, R7, R8, R>(f1:TernaryFunction<T1, T2, T3, R1>, f2:UnaryFunction<R1, R2>, f3:UnaryFunction<R2, R3>, f4:UnaryFunction<R3, R4>, f5:UnaryFunction<R4, R5>, f6:UnaryFunction<R5, R6>, f7:UnaryFunction<R6, R7>, f8:UnaryFunction<R7, R8>, f9:UnaryFunction<R8, R>): Pipeline<[T1, T2, T3], R>;
```

Comment:
Is there really no way of doing this without all the boilerplate? How about https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#generic-rest-parameters ?

Reply:
🤦‍♂️ of course ... well at least there is no need for the n-arity stuff. I thing the intermediate functions definition for `UnaryFunction` are (sadly) still necessary.

Updated the PR. Thanks for the feedback!